kernel/src/kernel.rs

```diff
@@ -104,13 +105,18 @@ impl Kernel {
         // index in the processes array. To avoid additional overhead, we do the
         // lookup and check here, rather than calling `.index()`.
         match self.processes.get(processid.index) {
-            Some(Some(process)) => {
-                // Check that the process stored here matches the identifier
-                // in the `processid`.
-                if process.processid() == processid {
-                    Some(*process)
-                } else {
-                    None
+            Some(pslot) => {
+                match pslot.get_active() {
+                    Some(process) => {
+                        // Check that the process stored here matches the identifier
+                        // in the `processid`.
+                        if process.processid() == processid {
+                            Some(process)
+                        } else {
+                            None
+                        }
+                    }
+                    None => None,
                 }
             }
             _ => None,
```

Comment:
I think this `match` could be written in a much simpler manner, something like:

```rust
self.processes.get(processid.index).map(|pslot| {
    match pslot.get_active() {
        // Check that the process stored here matches the identifier in
        // the `processid`.
        Some(process) if process.processid() == processid  => Some(process),
        _ => None,
    }
})
```

Or, avoiding `match` entirely (I prefer this, but it's somewhat subjective):

```rust
self.processes.get(processid.index)
              .and_then(|pslot| pslot.get_active())
              // Check that the process stored here matches the
              // identifier in the `processid`.
              .filter(|process| process.processid() == processid)
```